src/_pytest/compat.py

```diff
@@ -374,3 +385,29 @@ def overload(f):  # noqa: F811
     ATTRS_EQ_FIELD = "eq"
 else:
     ATTRS_EQ_FIELD = "cmp"
+
+
+# An implementation of functools.cached_property.
+# Can use the real one once only Python>=3.8 is supported.
+class cached_property(Generic[_S, _T]):
```

Comment:
Can we use this only under py<37, and on >=py36 we use `functools.cached_property` instead? Adding a check for Python version here also helps us drop this when the time comes.

Reply:
I didn't do it for these reasons: 

- the py38 implementation has some (minor) differences, namely it uses [`__set_name__`](https://www.python.org/dev/peps/pep-0487/), not available in py35, and is thread safe. So using it on >=py38 will introduce some "risky" variance.
- It is not implemented as a C extension or anything, so isn't any faster.
- Typing isn't ready yet (will be in the [next](https://github.com/python/typeshed/commit/add0b5e930a1db16560fde45a3b710eefc625709) mypy release).

I'll be happy to change it though if this doesn't convince you :)

Reply:
Thanks for the notes, appreciate them.

> the py38 implementation has some (minor) differences

At which point we can use the official version and drop our implementation, if they have diferences? Suppose this scenario (versions made up):

pytest 9.0 drops <py38, and drops the internal implementation in favor of the official one. py38, py39, py310 (the current versions supported) all of sudden act a little different because of those little differences.

Of course I'm just speculating here, but my gut feeling is to use the official implementation whenever available, so we can catch any problems due to small differences ASAP, and not all of sudden when we drop ours.

All in all, it probably it won't really mater as this is used only internally, but I wanted to share that.

Typing support might be an argument to not do the switch until the next `mypy` release happens though, if it affects us much.

Reply:
> At which point we can use the official version and drop our implementation, if they have diferences?

To use the official code directly, I think will require py36, because of `__set_name__`.

> Of course I'm just speculating here, but my gut feeling is to use the official implementation whenever available, so we can catch any problems due to small differences ASAP, and not all of sudden when we drop ours.

Makes sense. Updated to use the std version when available.